From 005ceaeee8b5e57232fef42a166832cfc77603dc Mon Sep 17 00:00:00 2001
From: Mark Stapp <mjs@cisco.com>
Date: Wed, 21 May 2025 15:43:08 -0400
Subject: [PATCH 1/5] lib,zebra: track the id of the nhgs used to resolve
 nexthops

When resolving recursive nexthops, capture the id of the nhg
used to resolve each nexthop. This will allow us to identify
the contribution the resolution processing makes to the
final "flattened" resolved nexthop.

Signed-off-by: Mark Stapp <mjs@cisco.com>
---
 lib/nexthop.c     | 1 +
 lib/nexthop.h     | 3 +++
 zebra/zebra_nhg.c | 6 ++++++
 3 files changed, 10 insertions(+)

diff --git a/lib/nexthop.c b/lib/nexthop.c
index 86e7b3c9ac..a8b30dc9b4 100644
--- a/lib/nexthop.c
+++ b/lib/nexthop.c
@@ -941,6 +941,7 @@ void nexthop_copy_no_recurse(struct nexthop *copy,
 		memcpy(copy->backup_idx, nexthop->backup_idx, copy->backup_num);
 
 	copy->srte_color = nexthop->srte_color;
+	copy->resolved_via = nexthop->resolved_via;
 	memcpy(&copy->gate, &nexthop->gate, sizeof(nexthop->gate));
 	memcpy(&copy->src, &nexthop->src, sizeof(nexthop->src));
 	memcpy(&copy->rmap_src, &nexthop->rmap_src, sizeof(nexthop->rmap_src));
diff --git a/lib/nexthop.h b/lib/nexthop.h
index 0a17b894f0..5af44983e7 100644
--- a/lib/nexthop.h
+++ b/lib/nexthop.h
@@ -159,6 +159,9 @@ struct nexthop {
 	/* SR-TE color used for matching SR-TE policies */
 	uint32_t srte_color;
 
+	/* If resolved (by zebra typically), the NHG ID of the resolver */
+	uint32_t resolved_via;
+
 	/* SRv6 information */
 	struct nexthop_srv6 *nh_srv6;
 };
diff --git a/zebra/zebra_nhg.c b/zebra/zebra_nhg.c
index c2bf266f39..90e7958a05 100644
--- a/zebra/zebra_nhg.c
+++ b/zebra/zebra_nhg.c
@@ -2674,6 +2674,12 @@ static int nexthop_active(struct nexthop *nexthop, struct nhg_hash_entry *nhe,
 				resolver = nexthop_set_resolved(afi, newhop, nexthop, NULL, flags);
 				resolved = 1;
 
+				/* Capture the NHG ID used to resolve the
+				 * nexthop.
+				 */
+				if (resolver)
+					resolver->resolved_via = match->nhe_id;
+
 				/* If there are backup nexthops, capture
 				 * that info with the resolving nexthop.
 				 */
-- 
2.50.1 (Apple Git-155)

